modules/auxiliary/scanner/rsync/modules_list.rb

```diff
@@ -6,64 +6,201 @@
+        false
+      else
+        vprint_error("#{peer} - unexpected response when connecting to #{rmodule}: #{res}")
+        'unknown'
+      end
+    else
+      vprint_error("#{peer} - no response when connecting to #{rmodule}")
+      'unknown'
+    end
+
+  end
+
+  def rsync_list
+    sock.puts("#list\n")
+
+    modules_metadata = []
+    # the module listing is the module name and comment separated by a tab, each module
+    # on its own line, lines separated with a newline
+    sock.get(read_timeout).split(/\n/).map(&:strip).map do |module_line|
+      break if module_line =~ /^#{RSYNC_HEADER} EXIT$/
+      name, comment = module_line.split(/\t/).map(&:strip)
+      next unless name
+      modules_metadata << { name: name, comment: comment }
+    end
+
+    modules_metadata
+  end
+
+  # Attempts to negotiate the rsync protocol with the endpoint.
+  def rsync_negotiate
+    # rsync is promiscuous and will send the negotitation and motd
+    # upon connecting.  abort if we get nothing
+    return unless (greeting = sock.get_once(-1, read_timeout))
+
+    # parse the greeting control and data lines.  With some systems, the data
+    # lines at this point will be the motd.
+    greeting_control_lines, greeting_data_lines = rsync_parse_lines(greeting)
+
+    # locate the rsync negotiation and complete it by just echo'ing
+    # back the same rsync version that it sent us
+    version = nil
+    greeting_control_lines.map do |greeting_control_line|
+      if /^#{RSYNC_HEADER} (?<version>\d+(\.\d+)?)$/ =~ greeting_control_line
+        version = Regexp.last_match('version')
+        sock.puts("#{RSYNC_HEADER} #{version}\n")
+      end
+    end
+
+    unless version
+      vprint_error("#{peer} - no rsync negotiation found")
+      return
+    end
+
+    _, post_neg_data_lines = rsync_parse_lines(sock.get_once(-1, read_timeout))
+    motd_lines = greeting_data_lines + post_neg_data_lines
+    [ version, motd_lines.empty? ? nil : motd_lines.join("\n") ]
+  end
+
+  # parses the control and data lines from the provided response data
+  def rsync_parse_lines(response_data)
+    control_lines = []
+    data_lines = []
+
+    if response_data
+      response_data.strip!
+      response_data.split(/\n/).map do |line|
+        if line =~ /^#{RSYNC_HEADER}/
+          control_lines << line
+        else
+          data_lines << line
+        end
+      end
+    end
+
+    [ control_lines, data_lines ]
   end
 
   def run_host(ip)
     connect
-    version = sock.get_once
-
-    return if version.blank?
-
-    print_good("#{ip}:#{rport} - rsync #{version.strip} found")
-    report_service(:host => ip, :port => rport, :proto => 'tcp', :name => 'rsync')
-    report_note(
-        :host => ip,
-        :proto => 'tcp',
-        :port => rport,
-        :type => 'rsync_version',
-        :data => version.strip
+    version, motd = rsync_negotiate
+    unless version
+      vprint_error("#{peer} - does not appear to be rsync")
+      disconnect
```

Comment:
It's good that you are disconnecting when needed, but the way it's done it looks like there's a resource leak. If there is a EOFError in rsync_negotiate (or other exceptions), then you may not get a chance to call #disconnect, which means the socket may not close properly. Please feel free to verify this. If this problem is true, then I think the best way to ensure #disconnec is put this in an ```ensure``` block. For example:

```
begin
  connect
  # do something. rsync_negotiate or party.
ensure
  disconnect
end
```

I have written a wiki before about using Msf::Exploit::Remote::Tcp. Please feel free to check it out:
https://github.com/rapid7/metasploit-framework/wiki/How-to-use-the-Msf%3A%3AExploit%3A%3ARemote%3A%3ATcp-mixin